### PR TITLE
Creating documentation with jsdocs

### LINKED ISSUE #5: Enhance documentation
The plugin needs some better documentation.  Could use jdocs to better describe the api.

### PR DESCRIPTION
**Pull request by urbantumbleweed**
#### What's this PR do?

Adds jsdocs documentation and adds a little more to the readme
#### Where should the reviewer start?

Take a look at the [README.md](README.md) read it, and then click on the _Docs_ link.
#### How should this be manually tested?

Just review content for clarity
#### Any background context you want to provide?

This is my first time using jsdocs and docdown.  Feedback on making this better is greatly appreciated.
#### What are the relevant tickets?

closes #5

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,17 @@
...
 Irken plugin for listening for key combinations
+
+####[Docs](/docs/iggins.md)
+
+#### How to use iggins:
+Register iggins on an Irken application.
+
+#### What iggins does:
+It exposes a `keypress` function that is used to register 'keyActions' on the Irken application. A 'keyAction' is comprised of:
+ * the string representation of a key combination (i.e., 'CTRL_SHIFT_L'), and
```
Comment (phated) [inline README.md]: this is incorrect, it is a predicate function, not a string and iggins exposes some common ones like `CTRL_SHIFT_L` on the object

File: README.md [hunk 2]
```diff
@@ -1,2 +1,17 @@
...
+ * the behavior(function) that should associated with the key event.
+
+Upon successfully registering a keyAction, the `keypress` function will return another function that can be used at a later time to remove newly registered 'keyAction' from the application.
+
+#### Who uses the returned function?
+The returned function is particularly important when the newly registered keyAction handler is bound to a DOM element that may at some future time cease to exist.  The returned function can be passed to the caller and invoked when the reference needs to be broken to allow for garbage collection.
```
Comment (phated) [inline README.md]: not bound, but interacts with

File: index.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ *   {
+ *     register: require('iggins')
+ *   }
+ * ];
+ *
+ * // Is registered on an Irken applicaion as a plugin.
+ * // This exposes the keypress functionality.
+ * app.register(plugins, function(err){
+ *  console.log('registered', err, app);
+ *  app.render(function(err){
```
Comment (phated) [inline index.js]: unnecessary for this example, examples should be the bare minimum

File: index.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+ *    console.log('rendered', err);
+ *  });
+ * });
+ *
+ * //For debugging
+ * window.app = app;
```
Comment (phated) [inline index.js]: unnecessary for this example, examples should be the bare minimum

File: index.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+ *
+ * //Verify that the keypress function is exposed on the Irken application.
+ * console.log(typeof app.keypress);
```
Comment (phated) [inline index.js]: registration is done async, this won't work or be available here, only inside the register callback

File: index.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+var _ = require('lodash');
+
+var predicates = require('./predicates');
+
+/**
+ * A plugin for Irken applications that exposes keyboard shortcut functionality.
```
Comment (phated) [inline index.js]: this is specifically the plugin registration function.  don't use the module description to describe the top level function (or any APIs)

File: index.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+ * @param  {object}   app  Instance of Irken application
+ * @param  {object}  opts Options to be passed to Iggins plugin.
+ * @param  {Function} done Noop function
```
Comment (phated) [inline index.js]: this isn't a noop, it is the callback to signal plugin setup is complete

File: index.js [hunk 8]
```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+var _ = require('lodash');
+
+var predicates = require('./predicates');
+
+/**
+ * A plugin for Irken applications that exposes keyboard shortcut functionality.
+ * @param  {object}   app  Instance of Irken application
+ * @param  {object}  opts Options to be passed to Iggins plugin.
```
Comment (phated) [inline index.js]: this is options passed during plugin registration

File: index.js [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+ * var plugins = [
+ *   {
+ *     register: require('iggins')
+ *   }
+ * ];
+ *
+ * // Is registered on an Irken applicaion as a plugin.
+ * // This exposes the keypress functionality.
+ * app.register(plugins, function(err){
+ *  console.log('registered', err, app);
```
Comment (phated) [inline index.js]: unnecessary for this example, examples should be the bare minimum

File: index.js [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+ * });
+ *
+ * //For debugging
+ * window.app = app;
+ *
+ * //Verify that the keypress function is exposed on the Irken application.
+ * console.log(typeof app.keypress);
+ * // => 'function'
+ */
+function iggins(app, opts, done) {
```
Comment (phated) [inline index.js]: not sure how I feel about documenting this as an API, as it isn't really one.  Plugin documentation is weird

**Comment by urbantumbleweed** [comment]
I updated the plugin to accept and handle `options.namespace`.  Descriptions for function params are revised.  The README is structured with install, usage, api reference, and api documentation including the option object.

**Comment by phated** [comment]
Other than the ordering, this is what I am able to produce with just jsdoc and docdown.

# index.js API documentation





## `iggins`
- <a href="#iggins">`iggins`</a>
- <a href="#iggins-keypress">`iggins.keypress`</a>





## `options`
- <a href="#options">`options`</a>
- <a href="#options-namespace">`options.namespace`</a>









## `iggins`



### <a id="iggins"></a>`iggins()`

<a href="#iggins">#</a> [&#x24C8;](https://github.com/iceddev/iggins/blob/master/src/index.js#L24) [&#x24C9;](#iggins)

Irken registration function

#### Example

``` js
function CTRL_F(evt) {
  return (evt.ctrlKey === true || evt.metaKey === true) && evt.keyCode === 70;
}
// => true or false

function handler(){
  console.log("Hello World!");
}

// registers a 'keyAction' on an application
var kp = iggins.keypress(CTRL_F, handler);
```

---





### <a id="iggins-keypress"></a>`iggins.keypress(predicate, handler)`

<a href="#iggins-keypress">#</a> [&#x24C8;](https://github.com/iceddev/iggins/blob/master/src/index.js#L53) [&#x24C9;](#iggins)

Registers a key combination predicate and handler as an application keyboard shortcut.

#### Arguments
1. `predicate` _(Function)_: The function used to match keydown events.
2. `handler` _(Function)_: The function to invoke when predicate returns true.

#### Returns

_(Function)_:            A keypress remover function to be called by requestor on DOM unmount.

---







## `options`



### <a id="options"></a>`options`

<a href="#options">#</a> [&#x24C8;](https://github.com/iceddev/iggins/blob/master/src/index.js#L33) [&#x24C9;](#iggins)

(Object): Options object passed during plugin registration.

---





### <a id="options-namespace"></a>`options.namespace`

<a href="#options-namespace">#</a> [&#x24C8;](https://github.com/iceddev/iggins/blob/master/src/index.js#L43) [&#x24C9;](#iggins)

(string): Namespace to expose.

---

**Comment by phated** [comment]
My stuff https://github.com/iceddev/iggins/tree/bleh

**Comment by urbantumbleweed** [comment]
I have made additions to predicates file and index.  I also updated the gulpfile so docs could be generated for all src files. Let me know if there is something else I need to do.  I thought about moving the README.md into src and the gulp-concating them to the README at root.  If you think it is a good idea, I will do it.

**Comment by phated** [comment]
Closing because I don't think JSDoc works for this use case.  Docs should be hand written in the README

**Pull request closed without merging**